### DISCUSSION TITLE
Are methods more than sugar?

### DISCUSSION THREAD

**Discussion by mindplay-dk** [vlang/v#15762] [General]
I was wondering, is there any real difference between methods and functions in V?

It doesn't look like V really has objects or method tables or anything like that? By the looks of it, methods are just functions with a different call syntax?

If so, why are methods even a thing?

Wouldn't it make more sense to just allow any function to be called using method call syntax for the first argument?

Then it would be just a choice if you use one syntax or the other - the functions would all just be functions with no special syntax in the declaration, and no constraints on the call syntax.

Is there a reason for methods beyond syntax?

**Comment by JalonSolov** [comment]
Because similar to class methods in OOP, V's methods are tied to the type.

Yes, of course you could use regular functions, but for example if you had a method that applied to strings, and a same-named method that applied to ints, to use regular functions instead you would either have to have a generic function that figured out which type it was called with, or you would have to have 2 differently named functions to handle the different types.

It really is just a different way of doing things, but it can lead to much cleaner syntax.

**Reply by Wajinn** [reply]
Prefer V's present way, because it does offer more clarity and readability.

**Comment by mindplay-dk** [comment]
I'm still wondering about this. 🤔

> Because similar to class methods in OOP, V's methods are tied to the type.

Are they?

According to the [manual](https://github.com/vlang/v/blob/master/doc/docs.md#methods):

> V doesn't have classes, but you can define methods on types. A method is a function with a special receiver argument. The receiver appears in its own argument list between the fn keyword and the method name. Methods must be in the same module as the receiver type.

And based on what you said here:

> ...to use regular functions instead you would either have to have a generic function that figured out which type it was called with, or you would have to have 2 differently named functions to handle the different types.

It looks to me as though V uses the type as part of the method's identity - but not really for anything else.

But that seems unnecessary? I mean, V is statically typed, so the type is attached to the symbol - it has the information it needs to resolve the call, doesn't it?

Let's take a practical example of potentially ambiguous methods: `array.reverse` and `string.reverse`.

```v
nums := [1, 2, 3]

println(nums.reverse()) // [3,2,1]

chars := "123"

println(chars.reverse()) // 321
```

What would be the problem with allowing the following?

```v
fn my_reverse(a []int) []int {
	return a.reverse()
}

fn my_reverse(s string) string {
	return s.reverse()
}

nums := [1, 2, 3]

println(nums.my_reverse()) // [3,2,1]
println(my_reverse(nums))  // [3,2,1]

chars := "123"

println(chars.my_reverse()) // 321
println(my_reverse(chars))  // 321
```

If this looks like overloading, it isn't - it's a semantic change, in which the first argument of *any* function becomes a part of the function identity, just as it does with the current function declaration syntax.

There's no ambiguity and no guesswork involved for the compiler here - this semantic change would make declarations like `fn f(x X) X` and `fn (x X) f() X` identical in meaning. (as such making the method declaration syntax obsolete.)

In terms of call semantics, "method calls" and "function calls" in V are both just function calls, right? As per the manual, "a method is a function with a special receiver argument". That's it, the call semantics are no different, I think?

The difference is how the functions are looked up at compile-time - method calls use the type and the function name, whereas function calls use only the function name.

The type + name identity approach is actually stronger. Why not apply it everywhere?

**Comment by MatejMagat305** [comment]
Well first g++ translated class into struct and method into functions so in the end every class is syntax sugar, in theorical it would be working equal, but in practical it would be adition work to compiler pass argument..., and it would be litle bit confusing fir programator now I can play to OOP and everybody from OOP will understand and they will be thinking that is OOP without knoledge backend and the last think in future will be able allow (in theoreticaly) play with visibility ...

**Comment by Wajinn** [comment]
> If so, why are methods even a thing?

The advantage of how V does it, is also the clarity and readability it provides.  The method is clearly assigned to a struct, even if the name of the methods are the same.  

If the person doesn't want to assign methods to structs, but wants parametric polymorphism, sum types and generics can be used with functions in V.

**Comment by mindplay-dk** [comment]
OOP Familiarity: yes, I don't think you lose that by allowing *even more* functions to be called as methods? And not everyone is looking for OOP - the qualities of functional programming definitely has more mind share today than it did even a few years ago. Unifying functions and methods would cater to both camps.

Clarity and readability: I think this argument is debatable. When you call a method defined in some other file, it's not immediately obvious where that method comes from or why it's callable. Explicit imports are arguably more ceremonious, but generally improve clarity and readability: if you're looking at just the one file, you're going to see, at the top of the file, where any functions come from. With good IDE support, adding import statements is easy, and the resulting source files can be read and understood more independently.

For what it's worth, [Nim has this feature](https://nim-lang.org/docs/manual.html#procedures-method-call-syntax) and this decision has worked out very well for the language - it has some [well-known minor limitations](https://nim-lang.org/docs/manual.html#templates-limitations-of-the-method-call-syntax), but (as argued) I think having to explicitly import your dependencies is "a good thing" for readability.

You might should take a look at Nim's codebase before dismissing the idea - in my opinion, it's very elegant, explicit, and readable* 🙂

###### * if you can see past the rest of Nim's syntax to this particular feature - arguably, much of the rest of the language isn't simple and looks more like a "systems programming language" than an "application language"...</small>

**Reply by Wajinn** [reply]
At some point and level, it can come down to preference.  So we may not be talking about what is better, but rather what we prefer.  It's like in some other threads where people preferred whitespace/off-side rule indentation to that of C-like languages.  Some prefer that style of indentation, while many others don't.  Furthermore, name dropping language preferences on the forums of competing languages usually doesn't go well.  It's like going to a C++ or Nim forum and telling them you prefer some other language or some feature from it, then get surprised there is push back or people think differently.  Though it's totally understandable to have a discussion, as a prelude to a feature request or just to see what others might think.

The other aspect about this, is that the way V does methods is not arbitrary, but is reflective of a more fundamental philosophy with a significantly long history, that can even be traced back to Nicklaus Wirth.  The creator of V mentions in the documentation that he was influenced by Go and Oberon.  If the matter is studied further, Oberon-2's way of doing methods directly influenced Go, and of which has strongly influenced V.  Consequently, V is configured more to be a descendant, alternative, and complimentary to Go and C.

**Reply by mindplay-dk** [reply]
I'm not "name dropping" Nim. I don't use it. I don't even have a preferred language.

I was just pointing at an interesting feature that might explain the concept better than I could.

I think this is a bit more important than bikeshedding tabs vs spaces.

Just trying to define and discuss a feature. Why the defensive attitude?

**Comment by medvednikov** [comment]
```
fn my_reverse(a []int) []int {
	return a.reverse()
}

fn my_reverse(s string) string {
	return s.reverse()
}
```

This would also require function overloading, which is a terrible idea. It slows down the compiler a lot, complicates the language, makes interop with other languages (primarily C) almost impossible, and makes large code bases much less readable.

V and Go have been designed to have no function overloading.

**Reply by mindplay-dk** [reply]
Take a closer look at [what I'm suggesting](https://github.com/vlang/v/discussions/15762#discussioncomment-4765303):

> If this looks like overloading, it isn't - it's a semantic change, in which **the first argument of any function becomes a part of the function identity**, just as it does with the current function declaration syntax.

Essentially, *all* functions would be methods in current sense of "methods" - so in your example, `fn my_reverse(a []int) []int` is a method of `[]int`, and `fn my_reverse(s string) string` is a method of `string`.

The function identity of *all* functions would be a tuple like `[first argument type, name]`.

The idea is to make functions and methods fully homogeneous - I think there is definitely something to be said for homogeneity in programming languages in general. "one concept, many purposes". It tends to make the language simpler. (for the author; not always for the language implementation.)

**Reply by InspColumbo** [reply]
So, if I got it right, in V methods are bound to data types by interfaces. It's actually like in OOP languages. Of course, V doesn't implement all the typical OOP functions with that alone.

If I understood your comment correctly, I think that a general alternative function call with a dot notation would only make sense in connection with a multiple dispatch approach, since the binding of the functions to the data types would then take place kind of automatically.

Personally, I think multiple dispatch is an excellent solution, but I got the impression that the V developers didn't implement this approach on purpose because I don't think that interfaces would be necessary otherwise.
V seems to me more like a mixture of the best things of GO and Rust with adding some good ideas but not with a revolution in mind.

Anyway, it could become one of the best programming languages so far.

**Reply by mindplay-dk** [reply]
I don't think this change requires multiple dispatch? Methods would still be statically resolved at compile-time, whether for method calls or interfaces - this change would affect *how* they get resolved, not *when*.

**Reply by InspColumbo** [reply]
Sorry, but I think multiple dispatch is exactly about how and not about when. As far as I know, it is called multiple dispatch when using all of the data types of a function's arguments and their order to choose which method should be invoked, rather than just the first.

**Reply by mindplay-dk** [reply]
I'm not proposing we use "all of the data types of a function's arguments" - only the first one, in exactly the same way we use the special first ("this") argument in methods now.

According to [this article](https://en.wikipedia.org/wiki/Multiple_dispatch#:~:text=Multiple%20dispatch%20or%20multimethods%20is,than%20one%20of%20its%20arguments.): (highlights mine)

> Multiple dispatch or multimethods is a feature of some programming languages in which a function or method can be **dynamically** dispatched based on the **run-time** (dynamic) type 

What I'm proposing does not rely on the run-time type.

> This is a generalization of single-dispatch polymorphism where a function or method call is dynamically dispatched based on the derived type of the object on which the method has been called. Multiple dispatch routes the dynamic dispatch to the implementing function or method using the combined characteristics of one or more arguments.

So from my understanding, I think "multiple" refers the the following:

> in languages with multiple dispatch, the selected method is simply the one whose arguments match the number and type of the function call. There is no special argument that owns the function/method carried out in a particular call.

If this *sounds* like overloading, also note the following:

> Multiple dispatch should be distinguished from [function overloading](https://en.wikipedia.org/wiki/Function_overloading), in which static typing information, such as a term's declared or inferred type (or base type in a language with subtyping) is used to determine which of several possibilities will be used at a given call site, and that determination is made at compile or link time (or some other time before program execution starts) and is thereafter invariant for a given deployment or run of the program.

Based on that, I don't think what I'm proposing is very different from what the language already does with methods: you can already have two methods with the same name - determination of the method is made at the call site based on (declared or inferred) static typing information.

With the change I'm proposing, the same determination would be made for any function based on it's first argument - I don't think that's different from what V does with methods already? Current syntax left, proposed semantic change right:

![image](https://user-images.githubusercontent.com/103348/223050033-9bb01922-30a6-4a14-ab25-01a8964c6c95.png)

Looks like the exact same amount of information to me - same information, statically determined, at the same time.

Instead of having two different kinds of functions, with different rules of determination, why not make them consistent?

Note that this isn't like extension methods in C# or something, if that's what you're worried about. With what I'm proposing, you would still need to explicitly import your functions. Extension methods in C# allow overloading - it uses the method name + the *entire* argument list as a (very weak, confusing, complex, and slow) form of method identity *globally*. V does not allow overloading.

I'm not a language engineer, so I could be wrong - but I don't think this change requires multiple dispatch?

**Reply by medvednikov** [reply]
Then how do you have methods with the same name, but for different objects?
```
user.delete()
repo.delete()

fn delete(u User) {}
fn delete(r Repo) {}
```

**Reply by mindplay-dk** [reply]
Same way you have them now? The type of `user` and `repo` are known and used to look up the appropriate method.

**Reply by InspColumbo** [reply]
Sorry, I still think you are wrong., mindplay, **...from Wikipedia:** Multiple dispatch or multimethods is a feature of some programming languages in which a function or method can be dynamically dispatched based on the run-time (dynamic) type **OR, in the more general case**, some other attribute of more than one of its arguments.

**Reply by medvednikov** [reply]
So you're proposing keeping the old syntax, but also add a way to do methods via functions. That's against V's one way rule.

**Reply by InspColumbo** [reply]
Alex, AFAIK, the type(objectname) can be added to the methodname during compilation.
E.g.,  methodname_typename or typename_methodname

when creating multiple dispatch functions, it's like

methodname_typename1_typename2_ typename3...

Each function would be unique during compilation then.
And Mindplay suggested just having an alternative writing for function calls...typename1.functionsname(typename2, typename3...)
Of course, V would make these things then totally different from Go but I think at the moment V is like a good evolution compared to other languges like Go and Rust. But if your language is the new kid on the block it should be a little bit more a revolution than a evolution to get the attention it needs.
Anyway, I love V so far and thank you for everything!

And sorry, English is not my mother language. I hope you can understand everything, though... :)

**Reply by InspColumbo** [reply]
To be honest, the 'one way rule' is good but maybe not always. :)
And at the moment methods in V are also called different from functions although they are only functions bound to data types.
There might be reasons where the dot notation is more readable and therefore OOP languages are often using it, I guess.
Also, you can consider forcing always writing the first argument in front of the functionname but that would brake the existing code what is not a good idea.
Anyway, using multiple dispatch in functions would make interfaces and methods obsolete in V what would make the language even more easy to use and would give it a big advantage over other languages I think.
And the usage of calling functions and 'object-functions' would look exactly like now. In fact, all functions would be bound to their parameters automatically.
I suggest allowing or forcing (forcing if you insist on the one way rule) the dot notation only when functions are using custom types as their first parameter which would look exactly like calling object methods. Anyway, the advantage would be that the definition of functions would always look the same. Just that simple.

At the moment I think the biggest advantage from Julia over Python is the performance and that Julia has multiple dispatch implemented. It's a dynamic scripting language and not comparable with V. I would really love to see such advantage from V over Go and other static compiled languages as well.

**Reply by Wajinn** [reply]
> Anyway, using multiple dispatch in functions would make interfaces and methods obsolete in V what would make the language even more easy to use and would give it a big advantage over other languages I think.

Such late changes would wreck V as we know it, and any supposed advantage seems more like conjecture and unproven beliefs.  It appears that one of the design goals of V is to be an alternative to Go (in addition to C), having a significant level of compatibility and evolving concepts from that branch (which stretch back to Oberon-2 and Wirth).  We see this (or attempts at it) with other newer languages, such as Nim compared to Python or Crystal with Ruby.  Such disruptive changes would derail that too, creating less attraction in its category, as it morphs V into something odd that has less clear objectives. 

> ...the definition of functions would always look the same...

This is mentioned, as a goal unto itself.  Looking the same, does not mean better or more useful.

**Reply by InspColumbo** [reply]
I can understand your arguments very well. Every design decision has its pros and cons, and it may be too late to make such fundamental changes, regardless of whether the programming language might attract more users or fewer as a result. Nevertheless, I think it is important to make such considerations so as not to miss opportunities. As said before, once the 1.0 release has happened, such changes can realistically no longer be made.
And sorry mindplay-dk for hijacking your thread for a short discussion. :)

**Reply by mindplay-dk** [reply]
> Sorry, I still think you are wrong., mindplay, **...from Wikipedia:** Multiple dispatch or multimethods is a feature of some programming languages in which a function or method can be dynamically dispatched based on the run-time (dynamic) type **OR, in the more general case**, some other attribute of more than one of its arguments.

"**more than one** of its arguments" - not what I'm proposing.

> So you're proposing keeping the old syntax, but also add a way to do methods via functions. That's against V's one way rule

I'm only suggesting you keep the old syntax for backwards compatibility - `v fmt` could handle the migration to a single syntax.

> Such late changes would wreck V as we know it

This won't wreck anything. You would run `v fmt` and everything would work as before.

> any supposed advantage seems more like conjecture and unproven beliefs

I don't think homogeneity in languages is a controversial principle? If one concept is enough to satisfy one requirement, then having one concept is *objectively* simpler than having two concepts for no practical reason.

In terms of superficial syntax, the supposed advantages are (1) all function declarations *look* the same, making the language easier to read and write, and (2) function or method style syntax can be chosen based on what makes the code easier to read.

The more subtle advantage is you effectively decouple the syntax from architectural decision making: you won't be tempted to choose a class just because chaining method calls *looks* nicer. If a method call returns a different type, you won't need to break your call chain, introduce extra variables, or wrap a call chain in a function call - you can break or continue call chains based on what is easier to read, or what best describes the problem/solution.

Which of the [following](https://play.vlang.io/p/eed9f111bb) more clearly describes the order in which the functions and methods are actually called?

1. `name.split(" ").map(it.runes()[0].str().to_upper() + it.runes()[1..].string().to_lower()).reverse().join(", ")`
2. `reverse(split(name, " ").map(to_upper(it.runes()[0].str()) + to_lower(it.runes()[1..].string()))).join(", ")`

The first form describes in a left-to-right reading order the steps that are actually applied - whereas the second form requires you to raise parens and scan back and forth to make any sense of it.

This is likely one reason why V itself prefers methods over functions for things like arrays and strings - but when you start adding your own functions that operate on strings or arrays (or any type of object outside your current library) you get some mix of (1) and (2) and readability suffers.

This change can fix that.

<details><summary>@medvednikov minor sidetrack</summary>
Without having looked at V's source code for string and array methods, I would guess at least some of these are not in fact methods, internally? Correct me if I'm wrong, but I would guess a string is not in fact an object, and you had to jump through some hoops to make the compiler "pretend" like strings or arrays actually have methods? If so, a semantic change might even simplify some of V's internal code.</details>

Again, I'm not a language engineer, and I don't know V's internal codebase, so I can't say if this change creates other challenges. But please don't dismiss the idea without giving it due consideration. It's quite a liberating feature, in my opinion - it removes a whole class of problems and considerations that most hybrid OOP/functional languages suffer from, so I think, by removing some mostly arbitrary architectural choices and decision making constraints from the developers mind, it aligns well with V's focus on simplicity.

(Of course barring it doesn't complicate things considerably for V's compiler internally, as I understand that simplicity is a priority both within the language and the language-implementation itself.)

**Reply by medvednikov** [reply]
> Correct me if I'm wrong, but I would guess a string is not in fact an object, and you had to jump through some hoops to make the compiler "pretend" like strings or arrays actually have methods? 

They are simple structs with simple methods:

https://github.com/vlang/v/blob/master/vlib/builtin/string.v#L44

> But please don't dismiss the idea without giving it due consideration.

we're always open to all ideas, and not dismissing; I have to do more research to understand this issue better

**Reply by InspColumbo** [reply]
To be honest, I think mindplay-dk has a very good and valid point with the readability of the dot notation.
However, I still think that if such a change, which would equate function and method calls, is actually implemented, the whole OOP concept in V would become obsolete as well.
I also think that this would give the programming language the decisive advantage over other languages. Just think about it. This would massively reduce the complexity, but the advantages would remain.

**Reply by mindplay-dk** [reply]
Well, V is already invested in OOP with features like access modifiers and interfaces - obsoleting OOP this late in the game would probably break too many things and might not be entirely realistic.

You don't necessarily have to unify functions and methods to address the syntax fluency issue either - there are at least two other options:

1. Just adding method-style function call syntax: leave methods as-is (offering struct access modifiers and interfaces) but allow functions to be called with the first argument. Similar to C# extension methods but minus the multiple dispatch. This would allow name collisions between methods and functions, so methods would probably need to "win" - the big drawback is, adding a new method could break existing code calling a function by the same name. So this isn't super elegant. But it's a much smaller change.

2. Adding some sort of [pipeline operator](https://github.com/tc39/proposal-pipeline-operator), another language feature adding a distinct "method-like" alternative calling syntax for functions. This avoids the problem with naming collisions. And it's definitely a much smaller and simpler feature - it's a small syntactic sugar intended solely to solve the syntax fluency problem. It's not a breaking change in any sense, and it's effects are entirely local to the expression.

And just my opinion on the mentioned pipeline operator proposal for JS: as you can see from the proposal, they want this operator with the additional `%` syntax to do more than just passing the left-hand side of the operator as the first argument. Personally, I don't think that's necessary - if the idea with this syntax is to improve readability, I'm not sure if these expression-like-but-not-quite-expressions do more harm or good. It's common for functions to accept the "subject" as the first argument anyway.

Arguably, both of these approaches *add* a bit of complexity (more semantics or more syntax) in favor of improving syntax fluency - whereas the idea of unifying functions and methods tries to *remove* some of the complexity from the language model.

Personally, if I was designing a language, I would definitely want the homogeneous functions and methods - but of course V already has an ecosystem and a userbase to consider. It would be awesome if there was some way to introduce this change progressively, but I also recognize this might not be easy - working around (preserving) access modifiers and interfaces is not something I can offer a solution to, not off the top of my head. Probably something a language engineer would need to weigh in on. 😊
